### ISSUE TITLE
Release in-memory log space based on entries size

### ISSUE THREAD

**Issue by pavelkalinnikov** [etcd-io/raft#71]
When entries exit the `unstable` log structure https://github.com/etcd-io/raft/blob/3e6cb625f59ca9da34d03df484217d874340e794/log_unstable.go#L156

it shrinks itself based on the entries slice length:
https://github.com/etcd-io/raft/blob/3e6cb625f59ca9da34d03df484217d874340e794/log_unstable.go#L162-L179

Firstly, the `len(u.entries)*lenMultiple < cap(u.entries)` does not necessarily do a right thing. After shifting the slice start in `u.entries = u.entries[num:]`, the first `num` entries "disappear" from `len` and `cap` of this slice: https://go.dev/play/p/ao1xaL0Edla.

So it's possible that `len >= cap/2` all the time, even though only a small portion of the backing slice is used. We should take the `cap` of the entire backing slice into account instead.

Secondly, this heuristic only takes `len/cap` of the slice into consideration, but not the size of the entries referenced by it. It could be beneficial to, in addition, shrink the slice if the sum size of  the "garbage" entries if more than a half. This would keep the overall memory consumption more controlled.

**Comment by pavelkalinnikov** [comment]
Alternatively, prove that the current strategy in `unstable` can't result in holding more memory than, say, 2x of a configured max.

**Comment by pavelkalinnikov** [comment]
But it doesn't look like we're safe as is. Say, the max size is 64 MB. Consider the following sequence of appends / truncations (numbers are entry sizes in MB):

```
2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 4 4 4 4 4 4 4 4 8 8 8 8 8 8 16 16 32
<-------------------------------> ....................................
                                  <-------------> ....................
                                                  <---------> ........
                                                              <---> ..
                                                                    <>
```

First, 17 entries of size 2MB are appended, total is 34 MB (below the limit). The slice `len == 17`, `cap == 32`.

Then, a sequence of 8 (append 4MB, truncate 2 x 2MB) follows. During this sequence, `len * 2 >= cap` holds, so we're not reallocating the array.

Then, a sequence of 4 (append 8MB, truncate 2 x 4MB) follows. Same thing.

Eventually, we fill up the `cap` of the array by a 32MB entry.

The total memory usage of this slice by the time we reallocate is 32MB * 5. More generally, an example can be constructed with O(B log B) bytes usage, for a configured max size B.

**Comment by CaojiamingAlan** [comment]
I am taking a look at this issue. 

However, from what I've learned, I don't think there is a way to get the length of entire backing array of a slice from the slice itself. That means we will also have to maintain  a `backingArrayLength` in the `unstable` struct, and update manually it everywhere we assign a value to `entries`. 

e.g. if we assign `entries = nil` then we set `backingArrayLength = 0`, and if we set `u.entries = append(u.entries, ents...)`, then we will have to check if the array is reallocated; if it is, then `backingArrayLength = cap(u.entries)`, else we leave it unchanged.

We need to do the same thing to calculate the accurate sum size of the "garbage" entries.

I wonder if I am correct and if this approach is expected.

BTW, there's a typo: there are 6 * 8MB entries in the figure. I think there should be only 4 of them.

**Comment by CaojiamingAlan** [comment]
I didn't see https://github.com/etcd-io/raft/issues/69. I think this issue would be easy to solve if we get https://github.com/etcd-io/raft/issues/69 done.

**Comment by Aditya-Sood** [comment]
hi @pavelkalinnikov @CaojiamingAlan 
is this issue still open for contribution?

**Comment by pavelkalinnikov** [comment]
> I wonder if I am correct and if this approach is expected.

@CaojiamingAlan Yes, I think you're correct, some kind of `backingArrayLength` / size tracking is needed to workaround. The `backingArrayLength` doesn't have to be baked into `unstable` type, you could think of the backing array as a separate "deque" kind of data structure and implement/test it separately to not overload `unstable` with the details.

@Aditya-Sood Still open, would you like to take a stab at it? @CaojiamingAlan Are you working on this?

**Comment by Aditya-Sood** [comment]
@pavelkalinnikov yes I'd like to try

if @CaojiamingAlan is unable to reply by Sunday I'll get started

**Comment by Aditya-Sood** [comment]
hi @pavelkalinnikov, I'll get started on this
could you assign the issue to me?

**Comment by beatrausch** [comment]
Hi,

not sure if you have heard about the new [protobuf editions](https://protobuf.dev/editions/overview/)

With editions golang got a new API called [Opaque](https://protobuf.dev/reference/go/go-generated-opaque/), which might help to move forward. Checkout the [Go Blog]( https://go.dev/blog/protobuf-opaque ) for some details.

Regards

**Comment by ahrtr** [comment]
This ticket is valid, but I'd like to keep the enhancement as simple as possible. Anyone feel free to continue to finish this.
```
$ git diff
diff --git a/log_unstable.go b/log_unstable.go
index f637b61..5dfa7fc 100644
--- a/log_unstable.go
+++ b/log_unstable.go
@@ -41,6 +41,12 @@ type unstable struct {
        entries []*pb.Entry
        // entries[i] has raft log position i+offset.
        offset uint64
+       // arrayCap is the capacity of the array backing entries, as of the last
+       // time that array was allocated. Unlike cap(entries), it is not reduced
+       // when stableTo slices entries off the front. This lets shrinkEntriesArray
+       // measure how much of the array is actually in use, even after many small
+       // calls to stableTo.
+       arrayCap int
 
        // if true, snapshot is being written to storage.
        snapshotInProgress bool
@@ -175,10 +181,12 @@ func (u *unstable) shrinkEntriesArray() {
        const lenMultiple = 2
        if len(u.entries) == 0 {
                u.entries = nil
-       } else if len(u.entries)*lenMultiple < cap(u.entries) {
+               u.arrayCap = 0
+       } else if len(u.entries)*lenMultiple < u.arrayCap {
                newEntries := make([]*pb.Entry, len(u.entries))
                copy(newEntries, u.entries)
                u.entries = newEntries
+               u.arrayCap = cap(newEntries)
        }
 }
 
@@ -193,6 +201,7 @@ func (u *unstable) restore(s *pb.Snapshot) {
        u.offset = s.GetMetadata().GetIndex() + 1
        u.offsetInProgress = u.offset
        u.entries = nil
+       u.arrayCap = 0
        u.snapshot = proto.Clone(s).(*pb.Snapshot)
        u.snapshotInProgress = false
 }
@@ -219,6 +228,12 @@ func (u *unstable) truncateAndAppend(ents []*pb.Entry) {
                // in-progress.
                u.offsetInProgress = min(u.offsetInProgress, fromIndex)
        }
+       // The above all either allocate a fresh array, or append starting from the
+       // same backing-array position as before, so cap(u.entries) is an accurate
+       // read of the whole backing array's capacity at this point. Record it so
+       // shrinkEntriesArray can still see it later, after stableTo has trimmed
+       // entries off the front and cap(u.entries) alone no longer reflects it.
+       u.arrayCap = cap(u.entries)
 }
 
 // slice returns the entries from the unstable log with indexes in the range
```

**Comment by ahrtr** [comment]
I re-read the change, the patch above has flaw. In `truncateAndAppend`,  the `u.arrayCap = cap(u.entries)` might not correct for the first case `case fromIndex == u.offset+uint64(len(u.entries))`, when the append reuses existing backing-array slack (no reallocation needed), `cap(u.entries)` reflects only physicalArraySize - currentStartOffset.

Also the following code (without the patch above) seems to be dead code. `shrinkEntriesArray` is only ever called from `stableTo`, right after `u.entries = u.entries[num:]`,  a front-trim. For any Go slice, slicing off the front by num decreases both `len` and `cap` by exactly `num` (cap of a sub-slice is always "distance from this slice's start to the physical end of the underlying array"). That means the slack S = `cap(entries) - len(entries)` is invariant under front-trimming — trimming can never change it.

https://github.com/etcd-io/raft/blob/621b0574681335100bb059b5cb3e2018070261cb/log_unstable.go#L178-L182

We can make the shrink more accurate, but it may complicate the implementation. It is unnecessary. In `truncateAndAppend`, it may automatically allocate a new slice. So I think we can just remove the dead code above. 

@MohanadKh03 sorry for the back and forth. Pls feel free to deliver a PR for that.

cc @pav-kv the initial reporter

**Comment by MohanadKh03** [comment]
> when the append reuses existing backing-array slack (no reallocation needed), cap(u.entries) reflects only physicalArraySize - currentStartOffset

After i added some changes with the `arrayCap` and tried to measure memory retained it seemed to stay the same which confused me but right it makes sense  

> We can make the shrink more accurate, but it may complicate the implementation. It is unnecessary.

i guess the only way is to measure not by len or cap alone but by the size of each entry , similar to how it's done in here right ? https://github.com/etcd-io/raft/blob/621b0574681335100bb059b5cb3e2018070261cb/log.go#L53-L60

I raised the PR [here](https://github.com/etcd-io/raft/pull/462) to remove the dead code but i guess this way shrinking as a whole could be removed too ? @ahrtr pls let me know if it would make sense to track backing array size and replace the current heuristic with it
